app/policies/wiki_policy.rb

```diff
@@ -1,3 +1,33 @@
 class WikiPolicy < ApplicationPolicy
+  class Scope
+    attr_reader :user, :scope
 
+    def initialize(user, scope)
+      @user = user
+      @scope = scope
+    end
+
+    def resolve
+      wikis = []
+      if user.role == 'admin'
+        wikis = scope.all
+      elsif user.role == 'premium'
+        all_wikis = scope.all
+        all_wikis.each do |wiki|
+          if !wiki.private || wiki.user_id == user.id
+            wikis << wiki
+          end
+        end
```

Comment:
This logic definitely works and is a good improvement over doing it in the view. However, you're still going and pulling all the wikis, and then iterating over all of them. That could get costly with a larger database. A preferable way to do this would be to use `ActiveRecord's where](http://guides.rubyonrails.org/active_record_querying.html#conditions) method to only retrieve the wikis you want. We don't spend a lot of time on SQL in the foundation, but I think it's definitely worth spending a couple hours playing with. This is a decentish intro to [WHERE](http://www.w3schools.com/sql/sql_where.asp). Let me know if you have any questions.